### PR TITLE
MODPATBLK-34 Limits for 3 implemented automated patron blocks not working properly

### PR DESCRIPTION
**Pull request by alexanderkurash**
Resolves [MODPATBLK-34](https://issues.folio.org/browse/MODPATBLK-34).

Changes blocking by 'Maximum number of items charged out' condition - blocks borrowing when the limit is reached. 
All other blocks are enabled only when a limit is exceeded.

### CODE REVIEW

File: src/main/java/org/folio/service/PatronBlocksService.java [hunk 1]
```diff
@@ -71,20 +74,32 @@ public PatronBlocksService(Map<String, String> okapiHeaders, Vertx vertx) {
...
 
     return CompositeFuture.all(new ArrayList<>(futures))
       .onFailure(log::error)
-      .onSuccess(cf -> blocks.getAutomatedPatronBlocks().addAll(cf.list()))
+      .onSuccess(cf -> blocks.getAutomatedPatronBlocks().addAll(
+        Arrays.stream(Condition.values())
+          .map(condition -> cf.list().stream()
+            .map(AutomatedPatronBlock.class::cast)
+            .filter(Objects::nonNull)
+            .filter(e -> e.getPatronBlockConditionId().equals(condition.getId()))
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/PatronBlocksService.java:87]: Please think of a better name for a variable

File: src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java [hunk 2]
```diff
@@ -110,240 +113,273 @@ public void shouldReturnNoBlocksWhenNoLimitsExistForPatronGroup() {
   }
 
   private void validateMaxNumberOfItemsChargedOutBlockResponse(int openLoansSizeDelta,
-    boolean singleLimit, boolean blockShouldExist) {
+    boolean singleLimit) {
 
     final Condition condition = MAX_NUMBER_OF_ITEMS_CHARGED_OUT;
+    updateCondition(condition, expectBlockBorrowing, expectBlockRenewals, expectBlockRequests);
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java:119]: Can we set all these to `true` by default before all tests?

File: src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java [hunk 3]
```diff
@@ -110,240 +113,273 @@ public void shouldReturnNoBlocksWhenNoLimitsExistForPatronGroup() {
...
   public void blockWhenMaxNumberOfItemsChargedOutLimitIsExceeded() {
-    validateMaxNumberOfItemsChargedOutBlockResponse(1, SINGLE_LIMIT, BLOCK_SHOULD_EXIST);
+    expectAllBlocks();
+    validateMaxNumberOfItemsChargedOutBlockResponse(1, SINGLE_LIMIT);
   }
 
   @Test
-  public void blockWhenMaxNumberOfItemsChargedOutLimitIsExceededAndAllLimitsExist() {
-    validateMaxNumberOfItemsChargedOutBlockResponse(1, ALL_LIMITS, BLOCK_SHOULD_EXIST);
+  public void blockWhenMaxNumberOfItemsChargedOutLimitIsExceededAndAllLimitsExist() {expectAllBlocks();
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java:167]: correct formatting

File: src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java [hunk 4]
```diff
@@ -79,6 +78,10 @@
     LIMIT_VALUES.put(MAX_OUTSTANDING_FEE_FINE_BALANCE, 70);
   }
 
+  private boolean expectBlockBorrowing = false;
+  private boolean expectBlockRenewals = false;
+  private boolean expectBlockRequests = false;
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java:83]: should we initialize these in `beforeEach()`?

File: src/main/java/org/folio/service/PatronBlocksService.java (hunk 1 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/service/PatronBlocksService.java]: This piece was removed

File: src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java (hunk 2 above)
Comment (alexanderkurash) [reply src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java]: Done

File: src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java (hunk 3 above)
Comment (alexanderkurash) [reply src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java]: Done

File: src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java (hunk 4 above)
Comment (alexanderkurash) [reply src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java]: Done

File: src/main/java/org/folio/domain/ActionBlocks.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+public class ActionBlocks {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private final boolean blockBorrowing;
+  private final boolean blockRenewals;
+  private final boolean blockRequests;
+
+  public static ActionBlocks byLimit(UserSummary userSummary, PatronBlockLimit limit) {
+
+    if (!ObjectUtils.allNotNull(userSummary, userSummary.getOpenLoans(), limit,
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/domain/ActionBlocks.java:32]: NPE if `userSummary` or `limit` is null

File: src/main/java/org/folio/service/PatronBlocksService.java [hunk 6]
```diff
@@ -50,59 +54,140 @@ public PatronBlocksService(Map<String, String> okapiHeaders, Vertx vertx) {
...
       .map(this::createBlockForLimit)
-      .collect(Collectors.toList());
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList()));
+
+    return blocks;
+  }
+
+  private Future<BlocksCalculationContext> addUserGroupIdToContext(BlocksCalculationContext ctx) {
+    if (!ObjectUtils.allNotNull(ctx.userSummary, ctx.userSummary.getUserId())) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/PatronBlocksService.java:90]: same here

File: src/main/java/org/folio/service/PatronBlocksService.java [hunk 7]
```diff
@@ -50,59 +54,140 @@ public PatronBlocksService(Map<String, String> okapiHeaders, Vertx vertx) {
...
 
-  private Future<AutomatedPatronBlock> createBlockForLimit(PatronBlockLimit limit) {
-    return succeededFuture(limit)
-      .compose(this::findConditionForLimit)
-      .map(this::createBlockForCondition);
+  private Future<BlocksCalculationContext> addPatronBlockLimitsToContext(
+    BlocksCalculationContext ctx) {
+
+    if (!ObjectUtils.allNotNull(ctx.userGroupId)) {
+      String message = "addPatronBlockLimitsToContext() failed - some ctx params are null";
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/PatronBlocksService.java:104]: Can we extract a builder method for this message with method name as parameter?

File: src/main/java/org/folio/service/PatronBlocksService.java [hunk 8]
```diff
@@ -50,59 +54,140 @@ public PatronBlocksService(Map<String, String> okapiHeaders, Vertx vertx) {
...
-        new EntityNotFoundException(format(
-          "Condition %s referenced by limit %s does not exist", conditionId, limit.getId()))));
+    return conditionsRepository.getAllWithDefaultLimit().map(ctx::withPatronBlockConditions);
   }
 
-  private AutomatedPatronBlock createBlockForCondition(PatronBlockCondition condition) {
+  private BlocksCalculationContext addCurrentConditionToContext(
+    BlocksCalculationContext ctx) {
+
+    if (!ObjectUtils.allNotNull(ctx.currentPatronBlockLimit,
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/PatronBlocksService.java:122]: same here

File: src/main/java/org/folio/service/PatronBlocksService.java [hunk 9]
```diff
@@ -50,59 +54,140 @@ public PatronBlocksService(Map<String, String> okapiHeaders, Vertx vertx) {
...
+
+    PatronBlockCondition patronBlockCondition = ctx.currentPatronBlockCondition;
+
+    ActionBlocks actionBlocksByLimit = ActionBlocks.byLimit(ctx.userSummary,
+      ctx.currentPatronBlockLimit);
+
+    ActionBlocks actionBlocksByCondition = new ActionBlocks(
+      Boolean.TRUE.equals(patronBlockCondition.getBlockBorrowing()),
+      Boolean.TRUE.equals(patronBlockCondition.getBlockRenewals()),
+      Boolean.TRUE.equals(patronBlockCondition.getBlockRequests()));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/service/PatronBlocksService.java:161]: NPE if we could not find condition by id during previous step.

**Pull request merged**